dimod/include/dimod/adjvectordqm.h

```diff
@@ -0,0 +1,356 @@
+        variable_type num_cases_v = this->num_cases(v);
+        for (variable_type case_v = 0; case_v < num_cases_v; case_v++) {
+            bqm_.set_linear(case_starts_[v] + case_v, biases[case_v]);
+        }
+    }
+
+    std::pair<bias_type, bool> get_quadratic_case(variable_type u,
+                                                  variable_type case_u,
+                                                  variable_type v,
+                                                  variable_type case_v) {
+        assert(u >= 0 && u < this->num_variables());
+        assert(case_u >= 0 && case_u < this->num_cases(u));
+        assert(v >= 0 && v < this->num_variables());
+        assert(case_v >= 0 && case_v < this->num_cases(v));
+        variable_type cu = case_starts_[u] + case_u;
+        variable_type cv = case_starts_[v] + case_v;
+        return bqm_.get_quadratic(cu, cv);
+    }
+
+    // Check if boolean type is still okay
+    bool set_quadratic_case(variable_type u, variable_type case_u,
+                            variable_type v, variable_type case_v,
+                            bias_type bias) {
+        assert(u >= 0 && u < this->num_variables());
+        assert(case_u >= 0 && case_u < this->num_cases(u));
+        assert(v >= 0 && v < this->num_variables());
+        assert(case_v >= 0 && case_v < this->num_cases(v));
+        if (u == v) {
+            return false;
+        }
+        variable_type cu = case_starts_[u] + case_u;
+        variable_type cv = case_starts_[v] + case_v;
+        bqm_.set_quadratic(cu, cv, bias);
+        connect_variables(u, v);
+        return true;
+    }
+
+    // Returns false if there is no interaction among the variables.
+    template <class io_bias_type>
+    bool get_quadratic(variable_type u, variable_type v,
+                       io_bias_type *quadratic_biases) {
+        assert(u >= 0 && u < this->num_variables());
+        assert(v >= 0 && v < this->num_variables());
+        if (!connection_present(u, v)) {
+            return false;
+        }
+        variable_type num_cases_u = num_cases(u);
+        variable_type num_cases_v = num_cases(v);
+        memset(
+            quadratic_biases, 0,
+            (size_t)num_cases_u * (size_t)num_cases_v * sizeof(io_bias_type));
+#pragma omp parallel for
+        for (variable_type case_u = 0; case_u < num_cases_u; case_u++) {
+            auto span =
+                bqm_.neighborhood(case_starts_[u] + case_u, case_starts_[v]);
+            while (span.first != span.second &&
+                   (span.first)->first < case_starts_[v + 1]) {
+                variable_type case_v = (span.first)->first - case_starts_[v];
+                quadratic_biases[case_u * num_cases_v + case_v] =
+                    (span.first)->second;
+                span.first++;
+            }
+        }
+        return true;
+    }
+
+    template <class io_bias_type>
+    bool set_quadratic(variable_type u, variable_type v, io_bias_type *biases) {
```

Comment:
I'd rather accept an iterator than a pointer (applies other places). Let me know if you disagree.

Reply:
If we pass in an iterator while passing a pointer we are sort of hiding the real information, we can always use the pointer as an iterator. The thing is we might end up doing pointer arithmetic which does not look too good on an iterator.

Reply:
I guess it seems to me that the iterator is the more general object. Specifically here we mean a random access iterator, since as you say we might end up doing arithmetic. But using an iterator instead would allow us to use strided arrays for instance.

Reply:
Let me think about it a bit, this is also coming up in the BQM case.